### PR TITLE
feat: callbacks -> async / await

### PR DESCRIPTION
**Pull request by alanshaw**
Branched from https://github.com/libp2p/js-libp2p-tcp/pull/102

TODO:

* [x] fix tests
* [x] ensure working as expected in interface-stream-muxer tests
* [x] fix interface-transport test to prevent flakey tests https://github.com/libp2p/interface-transport/pull/54

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,83 +1,135 @@
...
-      callback = options
-      options = {}
-    }
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Upgrader} options.upgrader
+   */
+  constructor ({ upgrader }) {
+    this._upgrader = upgrader
```
Comment (vasco-santos) [inline src/index.js]: should we `assert` that we receive the `upgrader`

Comment (jacobheun) [reply src/index.js]: Added the assert and a test. I included a link to the Upgrader definition in the interface-transport repo for UX.

File: src/index.js [hunk 2]
```diff
@@ -1,83 +1,137 @@
...
-      callback = options
-      options = {}
-    }
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Upgrader} options.upgrader
+   */
+  constructor ({ upgrader }) {
+    assert(upgrader, 'An upgrader must be provided. See https://github.com/libp2p/interface-transport#upgrader.')
```
Comment (achingbrain) [inline src/index.js]: The readme example has this class instantiable without any options.

Comment (jacobheun) [reply src/index.js]: Doing a PR shortly for another thing, I will include that there.

**Pull request merged**